python/python/lance/arrow.py

```diff
@@ -11,14 +11,347 @@
 #  limitations under the License.
-from typing import Optional, Union
+from pathlib import Path
+from typing import Iterable, Optional, Union
 
 import pyarrow as pa
 
 from .lance import BFloat16
 from .lance import bfloat16_array as bfloat16_array
 
 
+# TODO: add VariableShapeImageTensorType once pa.VariableShapeTensorArray is available
+class ImageURIType(pa.ExtensionType):
+    def __init__(self):
+        pa.ExtensionType.__init__(self, pa.string(), "lance.arrow.image_uri")
+
+    def __arrow_ext_serialize__(self):
+        return b""
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls, self, storage_type, serialized):
+        return ImageURIType()
+
+    def __arrow_ext_class__(self):
+        return ImageURIArray
+
+    def __arrow_ext_scalar_class__(self):
+        return ImageURIScalar
+
+
+class EncodedImageType(pa.ExtensionType):
+    def __init__(self):
+        # TODO: use pa.BinaryView once available?
+        pa.ExtensionType.__init__(self, pa.binary(), "lance.arrow.encoded_image")
+
+    def __arrow_ext_serialize__(self):
+        return b""
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls, self, storage_type, serialized):
+        return EncodedImageType()
+
+    def __arrow_ext_class__(self):
+        return EncodedImageArray
+
+    def __arrow_ext_scalar_class__(self):
+        return EncodedImageScalar
+
+
+class FixedShapeImageTensorType(pa.ExtensionType):
+    def __init__(self, arrow_type, shape):
+        pa.ExtensionType.__init__(
+            self,
+            pa.fixed_shape_tensor(arrow_type, shape),
+            "lance.arrow.fixed_shape_image_tensor",
+        )
+
+    def __arrow_ext_serialize__(self):
+        return b""
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls, self, storage_type, serialized):
+        return FixedShapeImageTensorType()
+
+    def __arrow_ext_class__(self):
+        return FixedShapeImageTensorArray
+
+    def __arrow_ext_scalar_class__(self):
+        return FixedShapeImageTensorScalar
+
+
+class ImageURIArray(pa.ExtensionArray):
+    def __repr__(self):
+        return "<lance.arrow.ImageURIArray object at 0x%016x>\n%s" % (
+            id(self),
+            repr(self.to_pylist()),
+        )
+
+    @classmethod
+    def from_uris(cls, uris: Union[pa.StringArray, Iterable[Union[str, Path]]]):
+        """
+        Create an ImageURIArray from a pa.StringArray or an iterable.
+
+        Parameters
+        ----------
+        uris : Union[pa.StringArray, Iterable[Union[str, Path]]]
+
+        Returns
+        -------
+        ImageURIArray
+            Array of image URIs
+
+        Examples
+        --------
+        >>> uris = ["file::///tmp/1.png"]
+        >>> ImageURIArray.from_uris(uris)
+        <lance.arrow.ImageURIArray object at 0x...>
+        ['file::///tmp/1.png']
+        """
+        from urllib.parse import urlparse
+
+        if isinstance(uris, Iterable):
+            uris = pa.array((str(uri) for uri in uris), type=pa.string())
```

Comment:
nit: if they are already passing a string array, can we just reuse that Arrow array instead of reconstructing it?

Reply:
Good catch. Switched to:
```python
if not isinstance(uris, pa.StringArray):
    uris = pa.array((str(uri) for uri in uris), type=pa.string())
```